packages/eslint-plugin/tools/generate-types.ts

```diff
@@ -0,0 +1,194 @@
+```
+
+Ultimately though we don't need 110% deep and correct types for our module because
+the types of the rules don't matter externally.
+
+This script generates approximate type declarations for the plugin -
+meaning that the keys will be correct, and the types of the values will be
+the supertype of the real values.
+
+This is enough for anyone who is consuming this plugin via types to use safely.
+
+Additional Notes:
+- We could manually define a `.d.ts` file for the module - but that's a pain to
+  maintain over time because we need to keep it in sync as we add/remove rules
+  and configs.
+- We can't use `tsc --noEmitOnError=false` for this because TS will not emit
+  `.d.ts` files for files with errors - only `.js` files.
+- We can't use `api-extractor` for this because that tool *only* operates on
+  `.d.ts` files - and we can't generate `.d.ts` files.
+*/
+
+// eslint-disable-next-line @typescript-eslint/internal/no-typescript-estree-import
+import {
+  AST_NODE_TYPES,
+  parse,
+  simpleTraverse,
+} from '@typescript-eslint/typescript-estree';
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as fs from 'fs';
+import makeDir from 'make-dir';
+import * as path from 'path';
+import prettier from 'prettier';
+
+const OUTPUT_PATH = path.resolve(__dirname, '..', 'dist', '_types');
+async function main(): Promise<void> {
+  await makeDir(OUTPUT_PATH);
+
+  const declaredNames = await Promise.all([
+    writeTypeDef('configs', {
+      importedName: 'Linter',
+      referenceName: 'Linter.Config',
+      moduleName: '@typescript-eslint/utils/ts-eslint',
+    }),
+    writeTypeDef('rules', {
+      importedName: 'RuleModule',
+      referenceName: 'RuleModule<string, readonly unknown[]>',
+      moduleName: '@typescript-eslint/utils/ts-eslint',
+    }),
+  ]);
+
+  // write the index file
+  const code = format([
+    ...declaredNames.map(name => `import type { ${name} } from './${name}'`),
+    '',
+    'declare const cjsExport: {',
+    ...declaredNames.map(name => `${name}: typeof ${name},`),
+    '};',
+    'export = cjsExport;',
+  ]);
+  await fs.promises.writeFile(
+    path.resolve(OUTPUT_PATH, 'index.d.ts'),
+    code,
+    'utf8',
+  );
+}
+
+async function writeTypeDef(
+  name: 'configs' | 'rules',
+  type: {
+    importedName: string;
+    referenceName: string;
+    moduleName: string;
+  },
+): Promise<string> {
+  const keys = await extractExportedKeys(
+    path.resolve(__dirname, '..', 'src', name, 'index.ts'),
+    name,
+  );
+
+  const code = format([
+    `import type { ${type.importedName} } from '${type.moduleName}';`,
+    '',
+    `type Keys = ${keys.map(k => `'${k}'`).join('\n  | ')};`,
+    '',
+    `export const ${name}: Readonly<Record<Keys, ${type.referenceName}>>;`,
+  ]);
+
+  await fs.promises.writeFile(
+    path.resolve(OUTPUT_PATH, `${name}.d.ts`),
+    code,
+    'utf8',
+  );
+
+  return name;
+}
+
+const BREAK_TRAVERSE = Symbol();
+async function extractExportedKeys(
```

Comment:
So I read through this and I think it makes sense for the approach of manually traversing through the source files and reading in their declaration nodes... but that's a lot of work. Why not just import from the built output and use the runtime values?

```js
import fs from 'node:fs/promises';
import plugin from '../dist/index.js';

await fs.writeFile(
  './dist/types/rules.d.ts',
  `import type { RuleModule } from '@typescript-eslint/utils/ts-eslint';

type Keys =
${Object.keys(plugin.rules)
  .map(ruleName => `  | '${ruleName}'`)
  .join('\n')}

export const rules: Readonly<
  Record<Keys, RuleModule<string, readonly unknown[]>>
>;
`,
);
```

(that, but with Prettier formatting, `make-dir` first, etc.)